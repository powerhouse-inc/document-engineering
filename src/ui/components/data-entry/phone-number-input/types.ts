import type React from 'react'
import type { InputBaseProps, WithDifference } from '../../../../scalars/components/types.js'
import type { SelectProps } from '../select/types.js'

type PrefixOptionFormat = 'CodesOnly' | 'FlagsOnly' | 'FlagsAndCodes' | 'FlagsAndNumbers'

interface PhoneNumberInputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, keyof InputBaseProps<string> | 'onChange'>,
    InputBaseProps<string>,
    Omit<WithDifference<string>, 'diffMode'> {
  onChange?: (value: string) => void
  allowedCountries?: string[]
  excludedCountries?: string[]
  includeDependentAreas?: boolean
  prefixOptionFormat?: PrefixOptionFormat
  prefixProps?: Pick<SelectProps, 'placeholder' | 'searchable' | 'className' | 'contentClassName'>
}

export type { PhoneNumberInputProps }
