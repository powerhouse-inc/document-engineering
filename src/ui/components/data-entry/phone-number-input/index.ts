export { PhoneNumberInput } from './phone-number-input.js'
export type { PhoneNumberInputProps } from './types.js'
