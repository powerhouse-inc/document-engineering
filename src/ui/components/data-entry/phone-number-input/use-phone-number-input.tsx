import { useCallback, useEffect, useMemo, useState } from 'react'
import { CircleFlag } from 'react-circle-flags'
import countries, { type Countries } from 'world-countries'
import parsePhoneNumber, { type CountryCode, getCountryCallingCode } from 'libphonenumber-js'
import type { PhoneNumberInputProps } from './types.js'
import type { SelectOption } from '../select/types.js'

interface UsePhoneNumberInputProps {
  value: PhoneNumberInputProps['value']
  defaultValue: PhoneNumberInputProps['defaultValue']
  onChange: PhoneNumberInputProps['onChange']
  onKeyDown: PhoneNumberInputProps['onKeyDown']
  allowedCountries: PhoneNumberInputProps['allowedCountries']
  excludedCountries: PhoneNumberInputProps['excludedCountries']
  includeDependentAreas: PhoneNumberInputProps['includeDependentAreas']
  prefixOptionFormat: PhoneNumberInputProps['prefixOptionFormat']
}

const renderFlagIcon = (countryCode: string) => {
  return <CircleFlag className="size-4" countryCode={countryCode.toLowerCase()} height={16} />
}

const getCallingCode = (countryCode: string): string | null => {
  try {
    const callingCode = getCountryCallingCode(countryCode as CountryCode)
    return `+${callingCode}`
  } catch {
    return null
  }
}

const parsePhoneValue = (rawValue: string) => {
  if (rawValue === '') return null

  const sanitizedValue = `+${rawValue.replace(/\D/g, '')}`
  const parsedValue = parsePhoneNumber(sanitizedValue, { extract: false })

  if (parsedValue?.isPossible() && parsedValue.country) {
    const callingCode = `+${parsedValue.countryCallingCode}`
    const selectValue = `${callingCode}-${parsedValue.country}`
    const inputValue = parsedValue.nationalNumber
    return { selectValue, inputValue }
  }

  return null
}

export const usePhoneNumberInput = ({
  value,
  defaultValue,
  onChange,
  onKeyDown,
  allowedCountries,
  excludedCountries,
  includeDependentAreas = false,
  prefixOptionFormat = 'FlagsAndNumbers',
}: UsePhoneNumberInputProps) => {
  const [selectValue, setSelectValue] = useState(() => {
    const parsedValue = parsePhoneValue(value ?? defaultValue ?? '')
    return parsedValue?.selectValue ?? ''
  })

  const [inputValue, setInputValue] = useState(() => {
    const rawValue = value ?? defaultValue ?? ''
    const parsedValue = parsePhoneValue(rawValue)
    return parsedValue?.inputValue ?? rawValue.replace(/\D/g, '')
  })

  const options: SelectOption[] = useMemo(() => {
    const defaultOptions = (countries as unknown as Countries)
      .filter(
        (country) =>
          (includeDependentAreas ? true : country.independent) &&
          country.cca2 !== 'AQ' && // exclude Antarctica
          getCallingCode(country.cca2) !== null // only countries with calling codes
      )
      .map((country) => {
        const callingCode = getCallingCode(country.cca2)!
        const countryCode = country.cca2

        let label = ''
        let icon: (() => JSX.Element) | undefined = undefined

        switch (prefixOptionFormat) {
          case 'CodesOnly':
            label = countryCode
            break
          case 'FlagsOnly':
            label = ''
            icon = () => renderFlagIcon(countryCode)
            break
          case 'FlagsAndCodes':
            label = countryCode
            icon = () => renderFlagIcon(countryCode)
            break
          case 'FlagsAndNumbers':
          default:
            label = callingCode
            icon = () => renderFlagIcon(countryCode)
            break
        }

        return {
          value: `${callingCode}-${countryCode}`,
          label,
          icon,
        }
      })
      .sort((a, b) => {
        const aCode = parseInt(a.value.split('-')[0].replace('+', ''))
        const bCode = parseInt(b.value.split('-')[0].replace('+', ''))
        return aCode - bCode
      })

    const allowedCountriesArray = Array.isArray(allowedCountries) ? allowedCountries : undefined
    const excludedCountriesArray = Array.isArray(excludedCountries) ? excludedCountries : undefined

    const filteredOptions =
      allowedCountriesArray || excludedCountriesArray
        ? defaultOptions.filter((option) => {
            const countryCode = option.value.split('-')[1]
            const isAllowedCountry = allowedCountriesArray ? allowedCountriesArray.includes(countryCode) : true
            const isExcludedCountry = excludedCountriesArray ? !excludedCountriesArray.includes(countryCode) : true
            return isAllowedCountry && isExcludedCountry
          })
        : defaultOptions

    return filteredOptions
  }, [allowedCountries, excludedCountries, includeDependentAreas, prefixOptionFormat])

  const handleSelectOnChange = useCallback(
    (newSelectValue: string) => {
      setSelectValue(newSelectValue)
      setInputValue('')
      const callingCode = newSelectValue.split('-')[0]
      onChange?.(callingCode)
    },
    [onChange]
  )

  const handleInputOnChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const callingCode = selectValue.split('-')[0]
      const parsedValue = parsePhoneValue(`${callingCode}${e.target.value}`)
      if (parsedValue && options.some((o) => o.value === parsedValue.selectValue)) {
        setSelectValue(parsedValue.selectValue)
        setInputValue(parsedValue.inputValue)
        const callingCode = parsedValue.selectValue.split('-')[0]
        onChange?.(`${callingCode}${parsedValue.inputValue}`)
      } else {
        const newInputValue = e.target.value.replace(/\D/g, '')
        setInputValue(newInputValue)
        onChange?.(`${callingCode}${newInputValue}`)
      }
    },
    [options, selectValue, onChange]
  )

  const handleOnKeyDown = useCallback(
    (e: React.KeyboardEvent<HTMLInputElement>) => {
      const isCopyShortcut =
        (e.ctrlKey && (e.key === 'c' || e.key === 'C')) || (e.metaKey && (e.key === 'c' || e.key === 'C'))

      const isCutShortcut =
        (e.ctrlKey && (e.key === 'x' || e.key === 'X')) || (e.metaKey && (e.key === 'x' || e.key === 'X'))

      const isPasteShortcut =
        (e.ctrlKey && (e.key === 'v' || e.key === 'V')) || (e.metaKey && (e.key === 'v' || e.key === 'V'))

      const isSelectAllShortcut =
        (e.ctrlKey && (e.key === 'a' || e.key === 'A')) || (e.metaKey && (e.key === 'a' || e.key === 'A'))

      const allowed = new Set([
        'Backspace',
        'Delete',
        'ArrowLeft',
        'ArrowRight',
        'ArrowUp',
        'ArrowDown',
        'Home',
        'End',
        'Tab',
        'Enter',
        'Escape',
      ])

      const isDigit = /^[0-9]$/.test(e.key)

      if (isCopyShortcut || isCutShortcut || isPasteShortcut || isSelectAllShortcut || allowed.has(e.key) || isDigit) {
        onKeyDown?.(e)
        return
      }

      e.preventDefault()
      onKeyDown?.(e)
    },
    [onKeyDown]
  )

  useEffect(() => {
    if (value !== undefined) {
      if (value === '') {
        setSelectValue('')
        setInputValue('')
        return
      }

      const parsedValue = parsePhoneValue(value)
      if (parsedValue && options.some((o) => o.value === parsedValue.selectValue)) {
        setSelectValue(parsedValue.selectValue)
        setInputValue(parsedValue.inputValue)
      }
    }
  }, [value, options])

  return {
    options,
    selectValue,
    inputValue,
    handleSelectOnChange,
    handleInputOnChange,
    handleOnKeyDown,
  }
}
