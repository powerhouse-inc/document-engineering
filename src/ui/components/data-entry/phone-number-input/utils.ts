import parsePhoneNumber, {
  type CountryCode,
  formatIncompletePhoneNumber,
  getCountryCallingCode,
} from 'libphonenumber-js'

export const getCallingCode = (countryCode: string): string | null => {
  try {
    const callingCode = getCountryCallingCode(countryCode as CountryCode)
    return `+${callingCode}`
  } catch {
    return null
  }
}

export const parsePhoneValue = (rawValue: string) => {
  if (rawValue === '') return null

  const withPlusValue = `${rawValue.startsWith('+') ? '' : '+'}${rawValue}`
  const parsedValue = parsePhoneNumber(withPlusValue, { extract: false })

  if (parsedValue?.isPossible() && parsedValue.country) {
    const expectedValue = `+${parsedValue.countryCallingCode}${parsedValue.nationalNumber}`
    if (withPlusValue !== expectedValue) {
      return null // the number was corrected by the library, reject it
    }

    const callingCode = `+${parsedValue.countryCallingCode}`
    const selectValue = `${callingCode}-${parsedValue.country}`
    const inputValue = parsedValue.nationalNumber
    return { selectValue, inputValue }
  }

  return null
}

export const formatPhoneNumber = (callingCode: string, inputValue: string): string => {
  if (callingCode === '' || inputValue === '') {
    return ''
  }

  const phoneNumber = `${callingCode}${inputValue}`
  const formatted = formatIncompletePhoneNumber(phoneNumber)

  if (formatted !== '' && formatted.length > callingCode.length) {
    const nationalPart = formatted.substring(callingCode.length).trim()
    return nationalPart || inputValue
  }

  return inputValue
}

export const removePhoneNumberFormat = (formattedValue: string): string => {
  return formattedValue.replace(/\D/g, '')
}

export const countDigitsBeforeCursor = (formattedText: string, cursorPos: number): number => {
  let digitCount = 0
  for (let i = 0; i < cursorPos && i < formattedText.length; i++) {
    if (/\d/.test(formattedText[i])) {
      digitCount++
    }
  }
  return digitCount
}

export const mapCursorPositionToFormattedText = (
  unformattedValue: string,
  formattedValue: string,
  targetDigitPosition: number
): number => {
  if (targetDigitPosition === 0) {
    return 0
  }

  // cursor at or beyond the end of unformatted value
  if (targetDigitPosition >= unformattedValue.length) {
    return formattedValue.length
  }

  // find the position in formatted value that corresponds to the target digit position
  let digitCount = 0
  for (let i = 0; i < formattedValue.length; i++) {
    if (/\d/.test(formattedValue[i])) {
      digitCount++
      if (digitCount === targetDigitPosition) {
        return i + 1
      }
    }
  }

  return formattedValue.length
}
