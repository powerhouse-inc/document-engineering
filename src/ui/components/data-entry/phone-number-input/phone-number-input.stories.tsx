import type { Meta, StoryObj } from '@storybook/react'
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from '../../../../scalars/lib/storybook-arg-types.js'
import { PhoneNumberInput } from './phone-number-input.js'

/**
 * ## Phone number input component
 *
 * `PhoneNumberInput` is a specialized input component for handling phone numbers.
 *
 * > **Note:** This component does not have built-in validation. If you need built-in validation
 * > you can use the [PhoneNumberField](?path=/docs/scalars-phone-number-field--readme)
 * > component.
 */

const meta: Meta<typeof PhoneNumberInput> = {
  title: 'Data Entry/Phone Number Input',
  component: PhoneNumberInput,
  decorators: [
    (Story) => (
      <div style={{ width: '280px', margin: '1rem auto 0' }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: 'padded',
  },
  tags: ['autodocs'],
  argTypes: {
    ...getDefaultArgTypes(),
    ...PrebuiltArgTypes.placeholder,
    ...getValidationArgTypes({
      enabledArgTypes: {
        validators: false,
        showErrorOnBlur: false,
        showErrorOnChange: false,
      },
    }),

    allowedCountries: {
      control: 'object',
      description: 'List of allowed country codes',
      table: {
        type: { summary: 'string[]' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    excludedCountries: {
      control: 'object',
      description: 'List of country codes to exclude',
      table: {
        type: { summary: 'string[]' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    includeDependentAreas: {
      control: 'boolean',
      description: 'Whether to include dependent territories in country list',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'false' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    prefixOptionFormat: {
      control: 'radio',
      options: ['CodesOnly', 'FlagsOnly', 'FlagsAndCodes', 'FlagsAndNumbers'],
      description: 'How to display country options in the prefix dropdown',
      table: {
        type: { summary: '"CodesOnly" | "FlagsOnly" | "FlagsAndCodes" | "FlagsAndNumbers"' },
        defaultValue: { summary: '"FlagsAndNumbers"' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    ...PrebuiltArgTypes.viewMode,
    ...PrebuiltArgTypes.baseValue,
  },
  args: {
    name: 'phone-number-input',
    label: 'Phone number',
  },
} satisfies Meta<typeof PhoneNumberInput>

export default meta
type Story = StoryObj<typeof PhoneNumberInput>

export const Default: Story = {}

export const Filled: Story = {
  args: {
    defaultValue: '+14155552671',
  },
}
