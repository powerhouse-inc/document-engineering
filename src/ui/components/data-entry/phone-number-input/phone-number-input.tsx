import React from 'react'
import { Select } from '../select/index.js'
import { Input } from '../input/index.js'
import { usePhoneNumberInput } from './use-phone-number-input.js'
import { cn, useUniqueId } from '../../../../scalars/lib/index.js'
import { FormDescription } from '../../../../scalars/components/fragments/form-description/index.js'
import { FormGroup } from '../../../../scalars/components/fragments/form-group/index.js'
import { FormLabel } from '../../../../scalars/components/fragments/form-label/index.js'
import { FormMessageList } from '../../../../scalars/components/fragments/form-message/index.js'
import type { PhoneNumberInputProps } from './types.js'

const PhoneNumberInput = React.forwardRef<HTMLInputElement, PhoneNumberInputProps>(
  (
    {
      id: idProp,
      className,
      label,
      description,
      value,
      defaultValue,
      onChange,
      disabled,
      required,
      errors,
      warnings,
      allowedCountries,
      excludedCountries,
      includeDependentAreas,
      optionFormat = 'FlagsAndNumbers',
      selectProps,
      viewMode = 'edition',
      // TODO: implement diffs
      // baseValue,
      ...props
    },
    ref
  ) => {
    const prefix = useUniqueId()
    const id = idProp ?? `${prefix}-phone-number-input`

    const { options, selectValue, inputValue, handleSelectOnChange, handleInputOnChange } = usePhoneNumberInput({
      value,
      defaultValue,
      onChange,
      allowedCountries,
      excludedCountries,
      includeDependentAreas,
      optionFormat,
    })

    const hasWarning = Array.isArray(warnings) && warnings.length > 0
    const hasError = Array.isArray(errors) && errors.length > 0

    if (viewMode !== 'edition') {
      // TODO: implement PhoneNumberInputDiff
      return null
    }

    return (
      <FormGroup>
        {!!label && (
          <FormLabel htmlFor={id} disabled={disabled} hasError={hasError} required={required}>
            {label}
          </FormLabel>
        )}
        <div
          className={cn(
            'relative flex items-center h-9 bg-white rounded-md border border-gray-300',
            'focus-within:ring-1 focus-within:ring-offset-0 focus-within:ring-gray-900'
          )}
        >
          <Select
            options={options}
            selectionIcon="checkmark"
            selectionIconPosition="right"
            searchable={selectProps?.searchable ?? true}
            disabled={disabled}
            value={selectValue}
            onChange={(newValue) => {
              handleSelectOnChange(newValue as string)
            }}
            placeholder={selectProps?.placeholder}
            className={cn(
              'focus:ring-0 focus:ring-offset-0',
              'outline-none focus-visible:ring-0 focus-visible:ring-offset-0',
              'h-full border-none rounded-l-md rounded-r-none',
              'bg-transparent focus:bg-transparent hover:bg-transparent',
              selectProps?.className
            )}
            contentClassName={cn('w-30', selectProps?.contentClassName)}
          />
          <div className="flex items-center h-full border-l border-gray-300" />
          <Input
            ref={ref}
            id={id}
            value={inputValue}
            onChange={handleInputOnChange}
            disabled={disabled}
            aria-invalid={hasError}
            aria-label={!label ? 'Phone number input' : undefined}
            aria-required={required}
            type="tel"
            className={cn(
              'focus:ring-0 focus:ring-offset-0',
              'outline-none focus-visible:ring-0 focus-visible:ring-offset-0',
              'h-full border-none rounded-r-md rounded-l-none',
              'bg-transparent',
              className
            )}
            {...props}
          />
        </div>
        {!!description && <FormDescription>{description}</FormDescription>}
        {hasWarning && <FormMessageList messages={warnings} type="warning" />}
        {hasError && <FormMessageList messages={errors} type="error" />}
      </FormGroup>
    )
  }
)

PhoneNumberInput.displayName = 'PhoneNumberInput'

export { PhoneNumberInput }
