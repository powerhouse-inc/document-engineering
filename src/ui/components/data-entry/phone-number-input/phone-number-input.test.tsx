import { render, screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { describe, expect, it, vi } from 'vitest'
import { PhoneNumberInput } from './phone-number-input.js'

describe('PhoneNumberInput Component', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<PhoneNumberInput name="phone" label="Phone number input" />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render with label', () => {
    render(<PhoneNumberInput name="phone" label="Phone number input" />)
    expect(screen.getByText('Phone number input')).toBeInTheDocument()
  })

  it('should render with description', () => {
    render(<PhoneNumberInput name="phone" description="Phone number input description" />)
    expect(screen.getByText('Phone number input description')).toBeInTheDocument()
  })

  it('should handle disabled state', () => {
    render(<PhoneNumberInput name="phone" disabled />)
    expect(screen.getByRole('combobox')).toBeDisabled()
    expect(screen.getByRole('textbox')).toBeDisabled()
  })

  it('should display error messages', async () => {
    render(<PhoneNumberInput name="phone" errors={['Invalid phone number']} />)
    await waitFor(() => {
      expect(screen.getByText('Invalid phone number')).toBeInTheDocument()
    })
  })

  it('should display warning messages', () => {
    render(<PhoneNumberInput name="phone" warnings={['Please change your phone number']} />)
    expect(screen.getByText('Please change your phone number')).toBeInTheDocument()
  })

  it('should have correct ARIA attributes', async () => {
    render(<PhoneNumberInput name="phone" label="Phone number input" required errors={['Invalid phone number']} />)

    const input = screen.getByRole('textbox')
    expect(input).toHaveAttribute('aria-required', 'true')
    await waitFor(() => {
      expect(input).toHaveAttribute('aria-invalid', 'true')
    })
  })

  it('should handle input value changes', async () => {
    const user = userEvent.setup()
    const onChange = vi.fn()

    render(<PhoneNumberInput name="phone" label="Phone number input" onChange={onChange} />)

    const input = screen.getByRole('textbox')
    await user.type(input, '4155552671')
    expect(onChange).toHaveBeenCalledTimes(10)
    expect(input).toHaveValue('4155552671')
  })

  it('should not invoke onChange on mount when it has a defaultValue', () => {
    const onChange = vi.fn()
    render(<PhoneNumberInput name="phone" label="Phone number input" defaultValue="+14155552671" onChange={onChange} />)
    expect(onChange).not.toHaveBeenCalled()
  })
})
