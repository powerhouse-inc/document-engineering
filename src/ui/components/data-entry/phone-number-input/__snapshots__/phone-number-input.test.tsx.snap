// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PhoneNumberInput Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
      for=":r0:-phone-number-input"
    >
      Phone number
    </label>
    <div
      class="relative flex items-center h-9 bg-white rounded-md border border-gray-300 focus-within:ring-1 focus-within:ring-offset-0 focus-within:ring-gray-900"
    >
      <div
        class="flex flex-col gap-2"
      >
        <button
          aria-controls="radix-:r2:"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-invalid="false"
          aria-label="Select"
          class="gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none disabled:pointer-events-none [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 flex items-center justify-between px-3 py-2 dark:border-charcoal-700 dark:bg-charcoal-900 rounded-md border border-gray-300 hover:border-gray-300 dark:hover:border-charcoal-700 dark:hover:bg-charcoal-800 dark:focus:ring-charcoal-300 focus:outline-none focus:ring-gray-900 dark:focus-visible:ring-charcoal-300 focus-visible:ring-gray-900 select focus:ring-0 focus:ring-offset-0 outline-none focus-visible:ring-0 focus-visible:ring-offset-0 h-full border-none rounded-l-md rounded-r-none bg-transparent focus:bg-transparent hover:bg-transparent w-auto shrink-0"
          data-state="closed"
          id=":r1:-select"
          role="combobox"
          type="button"
        >
          <div
            class="mx-auto flex w-full items-center justify-between gap-2"
          >
            <span
              class="text-[14px] font-normal leading-5 truncate text-gray-600 dark:text-gray-500"
            >
              +1
            </span>
            <svg
              class="cursor-pointer text-gray-700 dark:text-gray-400"
              data-testid="icon-fallback"
              fill="currentcolor"
              style="width: 16px; height: 16px;"
              viewBox="0 0 15 16"
            >
              <path
                clip-rule="evenodd"
                d="M4.93185 5.93185C4.75611 6.10759 4.75611 6.39251 4.93185 6.56825C5.10759 6.74398 5.39251 6.74398 5.56825 6.56825L7.50005 4.63644L9.43185 6.56825C9.60759 6.74398 9.89251 6.74398 10.0682 6.56825C10.244 6.39251 10.244 6.10759 10.0682 5.93185L7.81825 3.68185C7.73386 3.59746 7.6194 3.55005 7.50005 3.55005C7.3807 3.55005 7.26624 3.59746 7.18185 3.68185L4.93185 5.93185ZM10.0682 10.0682C10.244 9.89251 10.244 9.60759 10.0682 9.43185C9.89251 9.25612 9.60759 9.25612 9.43185 9.43185L7.50005 11.3637L5.56825 9.43185C5.39251 9.25612 5.10759 9.25612 4.93185 9.43185C4.75612 9.60759 4.75612 9.89251 4.93185 10.0682L7.18185 12.3182C7.35759 12.494 7.64251 12.494 7.81825 12.3182L10.0682 10.0682Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        class="flex items-center h-full border-l border-gray-300"
      />
      <input
        aria-invalid="false"
        class="flex w-full rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-gray-900 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 focus:ring-0 focus:ring-offset-0 outline-none focus-visible:ring-0 focus-visible:ring-offset-0 h-full border-none rounded-r-md rounded-l-none bg-transparent flex-1"
        id=":r0:-phone-number-input"
        name="phone"
        placeholder="4155552671"
        type="tel"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
